rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,866 @@
+pub async fn get_metrics_aggregate(
+    State(state): State<AppState>,
+    Query(q): Query<AggregateQuery>,
+) -> Result<Response, StatusCode> {
+    let now = chrono::Utc::now();
+    let timestamp = now.to_rfc3339();
+
+    // Parse attribute list (comma-separated), trim whitespace, drop empties
+    let group_attrs: Vec<_> = q
+        .attrs
+        .as_deref()
+        .unwrap_or("")
+        .split(',')
+        .map(|s| s.trim())
+        .filter(|s| !s.is_empty())
+        .collect();
+
+    // Aggregate groups with or without reset
+    let groups = aggregate_metric_groups(&state.metrics_registry, q.reset, &group_attrs);
+
+    // Resolve format (default json)
+    let fmt = q.format.unwrap_or_default();
+
+    match fmt {
+        OutputFormat::Json => {
+            let response = MetricsWithMetadata {
+                timestamp,
+                metric_sets: groups_with_metadata(&groups),
+            };
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::JsonCompact => {
+            let response = AllMetrics {
+                timestamp,
+                metric_sets: groups_without_metadata(&groups),
+            };
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::LineProtocol => {
+            let body = agg_line_protocol_text(&groups, Some(now.timestamp_millis()));
+            let mut resp = body.into_response();
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+        OutputFormat::Prometheus => {
+            let body = agg_prometheus_text(&groups, Some(now.timestamp_millis()));
+            let mut resp = body.into_response();
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+    }
+}
+
+fn aggregate_metric_groups(
+    registry: &MetricsRegistryHandle,
+    reset: bool,
+    group_attrs: &[&str],
+) -> Vec<AggregateGroup> {
+    // Aggregation map keyed by (set name, sorted list of (attr, Option<val_string>))
+    type GroupKey = (String, Vec<(String, Option<String>)>);
+    let mut agg: HashMap<
+        GroupKey,
+        (
+            HashMap<String, AttributeValue>,
+            HashMap<String, u64>,
+            &'static MetricsDescriptor,
+        ),
+    > = HashMap::new();
+
+    let mut visit = |descriptor: &'static MetricsDescriptor,
+                     attributes: &dyn AttributeSetHandler,
+                     metrics_iter: MetricsIterator<'_>| {
+        // Build key attributes vector
+        let mut key_attrs: Vec<(String, Option<String>)> = Vec::new();
+        if !group_attrs.is_empty() {
+            let mut amap: HashMap<&str, String> = HashMap::new();
```

Comment:
We are creating a HashMap out of the `attributes` twice once for `amap` and once for `avals`. I think we could create just one HashMap and it use it both places. That way we can avoid the double allocation for HashMaps and double iteration of `attributes`. We could have the value for the single HashMap contain both `String` and `AttributeValue` so that both places and use it.